Hello Yaniv!

My name is Olga. I'm happy to review your project today. 

When I will see mistake at the first time, I will just point it out. I let you find it and fix it by yourself. I'm trying to prepare you to work as an Data Analyst. SO, at a real job, your team lead will do the same. But if you can't handle this task yet, I will give you a more accurate hint at the next iteration.

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:


<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Success. Everything is done succesfully.
</div>


<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Remarks. Some recommendations.
</div>


<div class="alert alert-block alert-danger">
    
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
    
Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>


You can answer me by using this: 

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment</b> <a class="tocSkip"></a>
    
Yaniv, thank you for sending your project. You've done a really good job on it!

While there's room for improvement, on the whole, your project is impressive good. You did a great job! 

<span class="text-danger">There are one little thing that need to be corrected in your project. Could you check my comment? </span>
    
Every issue with our code is a chance for us to learn something new.

Good luck! :)
    
</div>

<div class="alert alert-block alert-warning">
<b>For your future projects</b> <a class="tocSkip"></a>
        
- It would be better to import all libraries in a first cell. It help us to set up of working environment.

- It is very good that you use Markdown! Please, use it more clearly. We can use #, ##, ###. etc for title, heading, subtitle, steps, etc. It helps to make our contents more structured

</div>


<div class="alert alert-block alert-info">
<b>First time I worked with Jupiter, before the next project I will learn how to use # ## ## and etc.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Overallv reviewer's comment. V.2.</b> <a class="tocSkip"></a>
    
Hello Yaniv!  Your corrections look great, you've improved your work significantly!
    
I'm glad to say that your project has been accepted.
    
Keep up the good work, and good luck on the next sprint! :) 
    
</div>

# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building the **credit score** of a potential customer. The **credit score** is used to evaluate the ability of a potential borrower to repay their loan.



**Before Im going to load the data and start my analyzing work Im adding the project description & main Q I been ask to answer and fill the report on, on the TOP of my project notebook, so I can always look in to them and make sure that Im working only on the problems I been ask to solve, and keep in maind the stracture of the work I need to do in the process of finding answers to the Q.** 

**Project description:**

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children have an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.
Your report will be considered when building a credit score for a potential customer. A credit score is used to evaluate the ability of a potential borrower to repay their loan.

**Answer these questions:**

1. Is there a connection between having **kids** and repaying a loan on time?
2. Is there a connection between **marital status** and repaying a loan on time?
3. Is there a connection between **income level** and repaying a loan on time?
4. How do different loan **purposes** affect on-time loan repayment?

**plan of work**

1. Read the DATA & learn the general information.
2. Data preprocessing.
3. Categorizion data
4. Study & check for conclusion in the process DATA and the Q I been asked.
5. Final conclusion. 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Great start with detailed introduction! :) </div>

In [1]:
# Loading all the libraries
import pandas as pd
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.stem import SnowballStemmer
english_stemmer = SnowballStemmer('english')
from collections import Counter




In [2]:
# Load the data
credit_score = pd.read_csv('/datasets/credit_scoring_eng.csv')

**Im reading the file, and save it to credit_score (NAME) which I took from the Task main GOAL** 

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
It would be better to import libraries in a first cell and read datasets in the second. Imagine we need to add a library. It will be easier to restart one cell than whole project.

</div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

# 1 Data exploration

**Description of the data**
- `children` - the number of children in the family
- `days_employed` - work experience in days
- `dob_years` - client's age in years
- `education` - client's education
- `education_id` - education identifier
- `family_status` - marital status
- `family_status_id` - marital status identifier
- `gender` - gender of the client
- `income_type` - type of employment
- `debt` - was there any debt on loan repayment
- `total_income` - monthly income
- `purpose` - the purpose of obtaining a loan



In [3]:

credit_score.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Conclusions**

1. From calling the (.info) I can learn that this dataset have 21525 rows in 12 diffrent columns.
2. I can also notice that before any changes the data in the columns have multipile types
_**[float64(2), int64(5), object(5)].**_
3. I can already learn that there is missing values/ missing info in 2 diffrent columns **(days_employed , total_income)**, & The missing values in both of the columns is the some - indicate that the issue maybe related in both.*
4. From the name of the column (days_employed) we can assum the type seposed to be int and not float, I will need to look
into that and see if there is some type of problem in the data (maybe nagative values??).
5. The name in the column **[dob_years]** probably related to the costumer **AGE** and therfor I will probably change the name of the column so it will be easier to look and study.*
    

In [4]:
credit_score.head(5)

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [5]:
credit_score.tail(5)

children  days_employed  dob_years            education  education_id  \
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income               purpose  
21520     35966.698  housing transactions  
21521     24959.969     purchase of a car  
21522     14347.610              property  
21523     39054.888     buying my own car  
21524     13127.587          to buy a car

In [6]:
credit_score.sample(5)

children  days_employed  dob_years            education  education_id  \
16880         0   -3168.852713         41    Bachelor's Degree             0   
13745         0    -527.266053         48  secondary education             1   
11916         0            NaN         28  secondary education             1   
11584         1            NaN         53  Secondary Education             1   
3171          1   -5399.597660         35    bachelor's degree             0   

      family_status  family_status_id gender income_type  debt  total_income  \
16880       married                 0      M    employee     0     22903.600   
13745      divorced                 3      F    employee     0     34796.566   
11916       married                 0      M    employee     0           NaN   
11584       married                 0      F    employee     0           NaN   
3171        married                 0      M    employee     0     19541.958   

                            purpose  
16880      purchase of my own house  
13745      real estate transactions  
11916  construction of own property  
11584                      property  
3171            building a property

**Conclusions**

1. I printed 3 diffrent sections in the table to obbserve the information it containe. After printing **(.head() , .tail() and .sample()** from my data set I can learn some more issues:
2. In the column (days_employed) I have negative values, which not seposed to be.
3. I will need to furthere investegate in to the data and the issues I found till now.


In [7]:
credit_score.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Conclusions**

As I could learn from the .info table, I can confirm that the missing values are only in 2 column and they have the exect sum of missing value. I can assum thay connected and maybe there was a problem with reading excel file or maybe thay been left in peurpse. I will need to keep my investegation before delete/fill them and that becuse atleast _[total_income]_ columns is in the heart of the reserch I need to do in this Task, and I cant just change it before learning the meaning of the data.

In [8]:
credit_score.describe(include = 'all')

children  days_employed     dob_years            education  \
count   21525.000000   19351.000000  21525.000000                21525   
unique           NaN            NaN           NaN                   15   
top              NaN            NaN           NaN  secondary education   
freq             NaN            NaN           NaN                13750   
mean        0.538908   63046.497661     43.293380                  NaN   
std         1.381587  140827.311974     12.574584                  NaN   
min        -1.000000  -18388.949901      0.000000                  NaN   
25%         0.000000   -2747.423625     33.000000                  NaN   
50%         0.000000   -1203.369529     42.000000                  NaN   
75%         1.000000    -291.095954     53.000000                  NaN   
max        20.000000  401755.400475     75.000000                  NaN   

        education_id family_status  family_status_id gender income_type  \
count   21525.000000         21525      21525.000000  21525       21525   
unique           NaN             5               NaN      3           8   
top              NaN       married               NaN      F    employee   
freq             NaN         12380               NaN  14236       11119   
mean        0.817236           NaN          0.972544    NaN         NaN   
std         0.548138           NaN          1.420324    NaN         NaN   
min         0.000000           NaN          0.000000    NaN         NaN   
25%         1.000000           NaN          0.000000    NaN         NaN   
50%         1.000000           NaN          0.000000    NaN         NaN   
75%         1.000000           NaN          1.000000    NaN         NaN   
max         4.000000           NaN          4.000000    NaN         NaN   

                debt   total_income           purpose  
count   21525.000000   19351.000000             21525  
unique           NaN            NaN                38  
top              NaN            NaN  wedding ceremony  
freq             NaN            NaN               797  
mean        0.080883   26787.568355               NaN  
std         0.272661   16475.450632               NaN  
min         0.000000    3306.762000               NaN  
25%         0.000000   16488.504500               NaN  
50%         0.000000   23202.870000               NaN  
75%         0.000000   32549.611000               NaN  
max         1.000000  362496.645000               NaN

In [9]:
credit_score.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [10]:
credit_score.isnull().sum() / len(credit_score)

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

In [11]:
for i in credit_score:
    if credit_score[i].isnull().sum()>0:
        print(i)


days_employed
total_income


**Conclusions**

From my research so far I can learn that there is 2 coulmns with missing values & that around 10% of _(days_employed & total_income )_ are missing. also I can learn that they seem symmetric.
Days_employed look like a complete damage data, that maybe come from damaged excel sheets, or maybe manual typo _(missing values, negative age values and co)_.
Its look like it going to be a impossible task to restore all the values in the days_employed coulmn.
Alto days employed is not part of the Q I need to focus on, I can't just delete the missing value raws or replace them with 0 for exemple, becuse I do have to foucus on total income in this task and there may be a connection between days_employed and total_income. I will do some more investagating on the _days_employed_ column and only then I will start changing the Values in the column. Also I will look in to the _(children,family_status,income level,purpose)_ from the _(describe)_table becuse those are the most segnificent columns for my research. and then I could write my conclusions from this part of my work.

In [12]:
credit_score['days_employed'].value_counts()

-986.927316     1
-7026.359174    1
-4236.274243    1
-6620.396473    1
-1238.560080    1
               ..
-2849.351119    1
-5619.328204    1
-448.829898     1
-1687.038672    1
-582.538413     1
Name: days_employed, Length: 19351, dtype: int64

In [13]:
credit_score['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
building a property                         620
purchase of my own house                    620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

In [14]:
credit_score['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [15]:
credit_score['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [16]:
credit_score['total_income'].value_counts()

17312.717    2
31791.384    2
42413.096    2
54857.666    1
26935.722    1
            ..
48796.341    1
34774.610    1
15710.698    1
19232.334    1
9591.824     1
Name: total_income, Length: 19348, dtype: int64

**Conclusions**

1. approx 10% of data in columns **["days_employed", "income_type"]** is missing.
2. I need to change type of **[days_employed]** from floth to Int.
3. I need transform the strings(object) in the columns to : lower.
4. I need to categorize data and lemmatize **[purpose]**.
5. 2174 records are missing information in columns **[‘days_employed’, ’total_income’]** I did not find a pattern according to which these values could have been missed.
6. some information has the same meaning like family_status and family_status_id.
7. **[children]** column have negative values - min=-1  and also very high values  max=20.
8. **[days_employed]** column, max is 401755, impossible value.
9. **[dob_years]**- min=0 impossible value.

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
So, we have basic information about our dataset. Let's start!</div>

# 2 Data Transformation

1. **Data preprocessing**
2. **Removing duplicates**
3. **Fixing information**

In [17]:
credit_score.rename(columns={'dob_years':'age'}, inplace = True)

In [18]:
credit_score.head()

children  days_employed  age            education  education_id  \
0         1   -8437.673028   42    bachelor's degree             0   
1         1   -4024.803754   36  secondary education             1   
2         0   -5623.422610   33  Secondary Education             1   
3         3   -4124.747207   32  secondary education             1   
4         0  340266.072047   53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [19]:
credit_score.groupby('income_type').total_income.agg(['count','min', 'median', 'mean', 'max'])

count        min      median          mean  \
income_type                                                               
business                      4577   4592.450  27577.2720  32386.793835   
civil servant                 1312   4672.012  24071.6695  27343.729582   
employee                     10014   3418.824  22815.1035  25820.841683   
entrepreneur                     1  79866.103  79866.1030  79866.103000   
paternity / maternity leave      1   8612.661   8612.6610   8612.661000   
retiree                       3443   3306.762  18962.3180  21940.394503   
student                          1  15712.260  15712.2600  15712.260000   
unemployed                       2   9593.119  21014.3605  21014.360500   

                                    max  
income_type                              
business                     362496.645  
civil servant                145672.235  
employee                     276204.162  
entrepreneur                  79866.103  
paternity / maternity leave    8612.661  
retiree                      117616.523  
student                       15712.260  
unemployed                    32435.602

In [20]:
credit_score.groupby('income_type').total_income.agg(['median'])

median
income_type                            
business                     27577.2720
civil servant                24071.6695
employee                     22815.1035
entrepreneur                 79866.1030
paternity / maternity leave   8612.6610
retiree                      18962.3180
student                      15712.2600
unemployed                   21014.3605

In [21]:
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'retiree')),'total_income']=18962.3180
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'employee')),'total_income']=22815.1035
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'civil servant')),'total_income']=24071.6695
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'business')),'total_income']=27577.2720
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'entrepreneur')),'total_income']=79866.1030
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'paternity / maternity leave')),'total_income']=8612.6610
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'student')),'total_income']=15712.2600
credit_score.loc[((credit_score.total_income.isnull())&
                 (credit_score.income_type == 'unemployed')),'total_income']=21014.3605

In [22]:
credit_score['total_income'].describe()

count     21525.000000
mean      26436.051922
std       15686.986477
min        3306.762000
25%       17247.708000
50%       22815.103500
75%       31287.991000
max      362496.645000
Name: total_income, dtype: float64

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Look, we have many professions here. And each of them have a different median income. Could fill missing values by grouping and median? </div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
Now we have more real values! :)  </div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
Also we can say:
    
    credit_score['total_income'] = credit_score['total_income'].fillna(credit_score.groupby('income_type'])['total_income'].transform('median'))
</div>


In [23]:
(abs(credit_score['days_employed'])/365).describe()

count    19351.000000
mean       183.328024
std        380.906522
min          0.066141
25%          2.539751
50%          6.011563
75%         15.172281
max       1100.699727
Name: days_employed, dtype: float64

In [24]:
credit_score[credit_score['days_employed'] <0] ['days_employed'].count()

15906

In [25]:
credit_score[credit_score['days_employed'] >=0] ['days_employed'].apply(lambda x: x / 365).describe()

count    3445.000000
mean     1000.011808
std        57.739771
min       900.626632
25%       949.697024
50%      1000.584401
75%      1049.990258
max      1100.699727
Name: days_employed, dtype: float64

<div class="alert alert-block alert-danger">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
~We should make days positive. :)~ </div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
Well done! Also we can use abs() method.  </div>

In [26]:
credit_score['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [27]:
credit_score['children']=credit_score['children'].replace(20, 2)

In [28]:
credit_score['children']=credit_score['children'].replace(-1, 1)

In [29]:
credit_score['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Well done! </div>


In [30]:
credit_score=credit_score.astype({ 'total_income':'int'})

In [31]:
credit_score.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
age                 21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        21525 non-null int64
purpose             21525 non-null object
dtypes: float64(1), int64(6), object(5)
memory usage: 2.0+ MB


In [32]:
credit_score['purpose'] = credit_score['purpose'].str.lower()

In [33]:
credit_score['education'] = credit_score['education'].str.lower()

In [34]:
credit_score['income_type'] = credit_score['income_type'].str.lower()

In [35]:
credit_score['gender'] = credit_score['gender'].str.lower()

In [36]:
credit_score['age'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: age, dtype: int64

In [37]:
age_avg = credit_score['age'].loc[(credit_score['age'] > 0)].median()

In [38]:
credit_score['age'] = credit_score['age'].replace(0, age_avg)

In [39]:
credit_score['age'].value_counts()

35    617
43    614
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: age, dtype: int64

In [40]:
credit_score['gender'].value_counts()

f      14236
m       7288
xna        1
Name: gender, dtype: int64

In [41]:
credit_score.loc[credit_score['gender'] == 'xna', 'gender'] = 'f'

In [42]:
credit_score['gender'].value_counts()

f    14237
m     7288
Name: gender, dtype: int64

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
We should fill by f, because we havelower case in a columns. Could you change you code a little?  </div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
Yep! </div>

In [43]:
credit_score.isnull().sum()


children               0
days_employed       2174
age                    0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

**Conclusions**

1. **There is no missing values left.**
2. **My next work will be dealing with duplicate DATA.**

In [44]:
credit_score['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [45]:
credit_score['education'].nunique()

5

In [46]:
credit_score.shape

(21525, 12)

In [47]:
credit_score.duplicated().sum()

71

In [48]:
credit_score.duplicated().sum()/len(credit_score*100)

0.0032984901277584204

In [49]:
credit_score = credit_score.drop_duplicates().reset_index(drop=True)

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Right decision! </div>

In [50]:
credit_score.shape

(21454, 12)

In [51]:
credit_score['purpose'].value_counts().count()

38

In [52]:
credit_score['purpose']

0          purchase of the house
1                   car purchase
2          purchase of the house
3        supplementary education
4              to have a wedding
                  ...           
21449       housing transactions
21450          purchase of a car
21451                   property
21452          buying my own car
21453               to buy a car
Name: purpose, Length: 21454, dtype: object

In [53]:
def type_purpose(purpose):
    for word in purpose.split(' '):
        if 'car' == english_stemmer.stem(word):
           return 'buy a car'
        if ('educ' == english_stemmer.stem(word) or 'univers' == english_stemmer.stem(word)):   
           return 'education'
        if ('hous' == english_stemmer.stem(word) or 'properti' == english_stemmer.stem(word) or 'estat' == english_stemmer.stem(word)):
           return 'real estate'
        if 'wed' == english_stemmer.stem(word):      
           return 'wedding'
    
    return purpose

credit_score['purpose'] = credit_score['purpose'].apply(type_purpose)


In [54]:
credit_score['purpose'].value_counts()

real estate    10811
buy a car       4306
education       4013
wedding         2324
Name: purpose, dtype: int64

**Conclusions**

1. **There where duplicate in ['purpose','education'].**
2. **I ysed drop_duplictes & stemming method to process them.**

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Great work with purposes! </div>



# 3 Categorization of data

1. **purpose - Categorical**
2. **income_type - Categorical**
3. **family_status Categorical**
4. **education Categorical**


In [55]:
children_cat = credit_score[['children']]
children_cat = children_cat.drop_duplicates().reset_index(drop=True)

def is_children(children):
    if children == 0 :
        return 'No children'
    if children == 1 :
        return '1 KID'
    if children == 2 :
        return '2 KIDS'
    if children == 3 :
        return '3 KIDS'
    if children == 4 :
        return '4 KIDS'
    if children > 4 :
        return '+4 KIDS'
    return 'Have children'

# adding the string field 'is_children'
children_cat['is_children'] = children_cat['children'].apply(is_children)
children_cat['is_children'].value_counts()

No children    1
2 KIDS         1
4 KIDS         1
1 KID          1
+4 KIDS        1
3 KIDS         1
Name: is_children, dtype: int64

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
We should check each number of children. Clients with 1 child have one value of debts, clients with 2 children have another value, etc  </div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
Very good! </div>

In [56]:
debt_cat = credit_score[['debt']]
debt_cat = debt_cat.drop_duplicates().reset_index(drop=True)

def is_debt(debt):
    if debt == 0 :
        return 'no debt'
    return 'there is a debt'

debt_cat['is_debt'] = debt_cat['debt'].apply(is_debt)
debt_cat['is_debt'].value_counts()

no debt            1
there is a debt    1
Name: is_debt, dtype: int64

In [57]:
def group_income (total_income):
    if total_income <15000 :
        return 0
    if total_income <20000 :
        return 1
    if total_income <25000 :
        return 2
    if total_income <30000 :
        return 3
    if total_income <40000 :
        return 4
    return 5

credit_score['group_income'] = credit_score['total_income'].apply(group_income)

In [58]:
total_income_dict = credit_score[['group_income']]
total_income_dict = total_income_dict.drop_duplicates().reset_index(drop=True)

In [59]:
def name_group_income (group_income):
    if group_income == 0 :
        return 'LOW <15000'
    if group_income == 1 :
        return 'MEDIUM_LOW 15000 - 20000'
    if group_income == 2 :
        return 'MEDIUM 20000 - 25000'
    if group_income == 3 :
        return 'MEDIUM_HIGH 25000 - 30000'
    if group_income == 4 :
        return 'HIGH 30000 - 40000'
    return 'VERY_HIGH >40000'
       
total_income_dict['name_group_income'] = total_income_dict['group_income'].apply(name_group_income)

In [60]:

total_income_dict.sort_values(by='group_income', ascending = True)

group_income          name_group_income
5             0                 LOW <15000
1             1   MEDIUM_LOW 15000 - 20000
2             2       MEDIUM 20000 - 25000
3             3  MEDIUM_HIGH 25000 - 30000
4             4         HIGH 30000 - 40000
0             5           VERY_HIGH >40000

**Conclusions**

After categorization There where 3 new dictionaries ['children_dict', 'debt_cat' and 'total_income_dict'] created

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Well done! </div>

# 4 Ansering Q


**Is there a correlation between having kids and paying back on time?**

In [61]:
pivot_table_childrens = credit_score.pivot_table(
    index='debt', columns='children', aggfunc={'debt':['count']})
pivot_table_childrens

debt                                  
            count                                  
children        0       1       2      3     4    5
debt                                               
0         13028.0  4410.0  1926.0  303.0  37.0  9.0
1          1063.0   445.0   202.0   27.0   4.0  NaN

In [62]:
pivot_table_childrens.loc['defaulted %'] = pivot_table_childrens.loc[1]*100/(pivot_table_childrens.loc[0] + pivot_table_childrens.loc[1])
pivot_table_childrens

debt                                                   \
                    count                                                    
children                0            1            2           3          4   
debt                                                                         
0            13028.000000  4410.000000  1926.000000  303.000000  37.000000   
1             1063.000000   445.000000   202.000000   27.000000   4.000000   
defaulted %      7.543822     9.165808     9.492481    8.181818   9.756098   

                  
                  
children       5  
debt              
0            9.0  
1            NaN  
defaulted %  NaN

**Conclusion**

**After reviewing the DATA and analze it I can come to A conclusion that having no kids at all will most lakely increase the propability of paying the loan on time, and on the atoher Hand having kids change the ability of paying the loan on time by MAX 2%.**

<div class="alert alert-block alert-danger">
<s><b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Waiting all children here. :) </div>

<div class="alert alert-block alert-info">
<b>Fix.</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment.V.2.</b> <a class="tocSkip"></a>
    
It was expected. :)  </div>

**Is there a correlation between family status and paying back on time?**

In [63]:
marital_debt = credit_score[['family_status','debt','purpose']]
marital_debt = debt_cat.merge(marital_debt, on='debt', how='right')

In [64]:
md_pivot = marital_debt.pivot_table(index='family_status', columns='is_debt', values='purpose', aggfunc='count')
md_pivot['ratio'] = md_pivot.loc[:,'there is a debt']/( md_pivot.loc[:,'no debt'] + md_pivot.loc[:,'there is a debt'])
md_pivot.sort_values(by='ratio', ascending = True)

is_debt            no debt  there is a debt     ratio
family_status                                        
widow / widower        896               63  0.065693
divorced              1110               85  0.071130
married              11408              931  0.075452
civil partnership     3763              388  0.093471
unmarried             2536              274  0.097509

**Conclusion**

**After reviewing the DATA and analze it I can come to A conclusion that 'widow / widower' & 'divorced' have less overdue loans, and the 'unmarried' group has the worst loan repayment.** 

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Intresting information! :) </div>

**Is there a correlation between income level and paying back on time?**

In [65]:
total_income_debt = credit_score[['group_income','debt','purpose']]
total_income_debt = total_income_dict.merge(total_income_debt, on='group_income', how='right')
total_income_debt = debt_cat.merge(total_income_debt, on='debt', how='right')
total_income_debt

debt          is_debt  group_income name_group_income      purpose
0         0          no debt             5  VERY_HIGH >40000  real estate
1         0          no debt             5  VERY_HIGH >40000    education
2         0          no debt             5  VERY_HIGH >40000  real estate
3         0          no debt             5  VERY_HIGH >40000  real estate
4         0          no debt             5  VERY_HIGH >40000  real estate
...     ...              ...           ...               ...          ...
21449     1  there is a debt             0        LOW <15000  real estate
21450     1  there is a debt             0        LOW <15000    buy a car
21451     1  there is a debt             0        LOW <15000  real estate
21452     1  there is a debt             0        LOW <15000    buy a car
21453     1  there is a debt             0        LOW <15000  real estate

[21454 rows x 5 columns]

In [66]:
income_debt_pivot = total_income_debt.pivot_table(index='name_group_income', columns='is_debt', values='purpose', aggfunc='count')
income_debt_pivot['ratio'] = income_debt_pivot.loc[:,'there is a debt']/( income_debt_pivot.loc[:,'no debt'] + income_debt_pivot.loc[:,'there is a debt'])
income_debt_pivot.sort_values(by='ratio', ascending = True)

is_debt                    no debt  there is a debt     ratio
name_group_income                                            
VERY_HIGH >40000              2619              194  0.068966
HIGH 30000 - 40000            2865              242  0.077889
LOW <15000                    3445              298  0.079615
MEDIUM_HIGH 25000 - 30000     2916              269  0.084458
MEDIUM 20000 - 25000          4201              393  0.085546
MEDIUM_LOW 15000 - 20000      3667              345  0.085992

**Conclusion**

**After analyzing the data, i learn that the best borrowers are clients with an income of 40,000 or more & the most unreliable borrowers are clients with an income of 15,000**

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Not so expected.  :) </div>

**How does credit purpose affect the default rate?**

In [67]:
for cat in credit_score['purpose'].unique():
    purpose_category = credit_score['purpose'][(credit_score['purpose']==cat) & (credit_score['debt']==0)].count()
    total = credit_score['purpose'][credit_score['purpose']==cat].count()
    purpose_percent = round(purpose_category/ total * 100, 2)
    print(f"{purpose_percent}% {cat} purpose category have paid their loans on time")

92.77% real estate purpose category have paid their loans on time
90.64% buy a car purpose category have paid their loans on time
90.78% education purpose category have paid their loans on time
92.0% wedding purpose category have paid their loans on time


In [68]:
pivot_purpose = credit_score.pivot_table(index=['purpose'], values=['debt'], aggfunc=['sum', 'count'])
pivot_purpose['ratio'] = round(pivot_purpose[('sum', 'debt')]/ pivot_purpose['count', 'debt'] * 100, 2)
pivot_purpose

sum  count ratio
            debt   debt      
purpose                      
buy a car    403   4306  9.36
education    370   4013  9.22
real estate  782  10811  7.23
wedding      186   2324  8.00

**Conclusion**

**All purpose category have more than 90% probability to pay their loans on time.**
**Car and education have more missing payments than other purpose category.**

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
So, real estate purpose category has many bank check and it is the best.  :) </div>

# General Conclusion 

1. **Q1 CONCLUSION** = After reviewing the DATA and analze it I can come to A conclusion that having no kids at all will most lakely increase the propability of paying the loan on time, and on the atoher end having no kids dosent really change the ability of paying the loan on time.
2. **Q2 CONCLUSION** = After reviewing the DATA and analze it I can come to A conclusion that 'widow / widower' & 'divorced' have less overdue loans, and the 'unmarried' group has the worst loan repayment. 
3. **Q3 CONCLUSION** = After analyzing the data, i learn that the best borrowers are clients with an income of 40,000 or more & the most unreliable borrowers are clients with an income of 15,000
4. **Q4 CONCLUSION** = All purpose category have more than 90% probability to pay their loans on time. Car and education have more missing payments than other purpose category.

**The data contained some anomalies, including errors in data insertion and lack of categorization, making analysis difficult.
From this data set I wasn't able to discover a relation between on-time loan repayment and their cause. From the 4 analysis that I made the categories show that more than 90% of the members will repay their loan on time.
In general, no correlation has been identified. There is a weak correlation between having children, also purposes of loans could impact the picture of the credit score in general speaking. It can be assumed that there is correlation between multiple features at the same time. Also we could describe portrait of good client, - without kids, with total income more than 200000.
customers without children have a slightly lower default-rate than customers with 1 or 2 children
widows/widowers have a litle bit lower default-rate than all bank's customers in average
unmarried customers have a little bit higher default-rate than all customers in average
customers with a very high income (1% of customers with the highest income) have a litle bit lower default-rate than all customers in average
customers without real estate loan purpose have a slightly lower default-rate than customers with car or education loan purposes.
alto I didnt focus on that I could assume from looking in the table DATA that there is more connection by looking on age and income to the ability of returning a loan, but I didnt focus on that in this TASK.**

<div class="alert alert-block alert-info">
<b>THANK YOU VERY MUCH FOR REVIEWING MY WORK AND GIVING ME INPUTS AND POINTS, Y.T</b> <a class="tocSkip"></a>
</div>

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>
    
Very good final conclusion! Thank you for your work! </div>